internal/controller/object/object.go

```diff
@@ -312,6 +312,28 @@ func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
 	if observation, err = c.getObservation(cr, observed); err != nil {
 		return managed.ExternalObservation{}, errors.Wrap(err, errGetLastApplied)
 	}
+
+	if c.ssaExtractor != nil {
+		// Note(turkenh): This dry run call is mostly a workaround for the
+		// following issue: https://github.com/kubernetes/kubernetes/issues/115563
+		// In an ideal world, we should be able to compare the extracted observation
+		// with what we will apply as desired state, however, due to the poor
+		// handling of defaults in the server-side apply, we cannot do that,
+		// since we always see a diff between the extracted state and the desired
+		// state in that case. This dry run call returns what we will see on
+		// the object including the defaults at the cost of one extra call to
+		// the apiserver, so that we can compare it with the extracted state to
+		// decide whether the object is up-to-date or not.
+		desiredObj := desired.DeepCopy()
+		if err := c.client.Patch(ctx, desiredObj, client.Apply, client.ForceOwnership, client.FieldOwner(ssaFieldOwner(cr.Name)), client.DryRunAll); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(CleanErr(err), "cannot dry run SSA")
+		}
+		desired, err = c.ssaExtractor.Extract(desiredObj, ssaFieldOwner(cr.Name))
+		if err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, "cannot extract SSA")
+		}
+	}
+
```

Comment:
> re-apply our desired state on every event

That wouldn't work since, in some cases, re-applying the same manifest simply [bumps the resourceVersion](https://github.com/kubernetes/kubernetes/issues/124605) and causing another update event. So, if we implement it in that way, we would end up in an update loop, i.e., apply -> update event -> apply -> update event ...

There are various issues that I believe all boil down to the same root cause:

- https://github.com/kubernetes/kubernetes/issues/124605
- https://github.com/kubernetes/kubernetes/issues/121404
- https://github.com/kubernetes/kubernetes/issues/124050
- https://github.com/kubernetes/kubernetes/issues/115563

> If we have a apiserver roundtrip anyway for dry-run

So, if the above bug didn't exist, we wouldn't need that dry-run. This is already a workaround for that.

Reply:
> So, if we implement it in that way, we would end up in an update loop, i.e., apply -> update event -> apply -> update event

Given that composition functions uses SSA to apply composed resources, should we expect the same problem in c/c if realtime compositions are used together with functions?

Reply:
Probably. 

I reproduced with kubectl using [the deployment in this issue](https://github.com/kubernetes/kubernetes/issues/124050), and know that an empty map triggers it. So, it doesn't happen for all resources but for some. I am almost sure there will be some MRs resulting in similar behavior under certain conditions. [This](https://github.com/kubernetes/kubernetes/issues/124605) is reproducing with a CRD / CR for example.

Reply:
That's quite unfortunate. 🙁 I've added some notes to the SSA and realtime composition lifecycle tracking issues just in case we see this in the wild.